Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp

```diff
@@ -82,17 +82,26 @@ void CoordinatedBackingStore::paintToTextureMapper(TextureMapper& textureMapper,
                 && !tile.texture().filterOperation();
         }();
         if (canUseDamageToDrawTextureFragment) {
-            // We define damagedTileRect as a minimum bounding rectangle of all damage rects that intersect tile.rect()
-            // - this way we can keep a single texture draw call yet with potentially smaller sourceRect.
-            FloatRect damagedTileRect;
+            // We use Damage here as it's a convenient way to track per-tile damage.
+            // If the tile size is big enough, we use Damage::Mode::Rectangles so that
+            // more than one "draw" operation can be performed thus reducing the re-painted
+            // area significantly in many cases.
+            const bool shouldUseAtMostOneDrawCall = tile.rect().width() <= Damage::s_cellSize && tile.rect().height() <= Damage::s_cellSize;
+            IntRect tileIntRect = enclosingIntRect(tile.rect());
+            Damage tileDamage(tileIntRect, shouldUseAtMostOneDrawCall ? Damage::Mode::BoundingBox : Damage::Mode::Rectangles);
             for (const auto& damageRect : frameDamage->rects()) {
-                if (!damageRect.isEmpty())
-                    damagedTileRect.unite(intersection(tile.rect(), damageRect));
+                if (damageRect.isEmpty())
+                    continue;
+                IntRect tileDamageRect = intersection(tileIntRect, damageRect);
+                if (tileDamageRect.isEmpty())
+                    continue;
+                tileDamage.add(tileDamageRect);
```

Comment:
Damage::add already returns early if the rectangle is empty, so this could be simplified and just do:

```
tileDamage.add(intersection(tileIntRect, damageRect));
```